src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs

```diff
@@ -254,30 +258,47 @@ private unsafe bool HandleItems(ActionType actionType, QueryOperationContext que
                                         }
                                     }
 
-                                    if (CanContinueBatch(queryContext, indexContext, collectionStats, indexWriter, lastEtag, lastCollectionEtag, totalProcessedCount) == false)
-                                    {
-                                        keepRunning = false;
+                                    if (earlyExit)
                                         break;
-                                    }
 
-                                    if (totalProcessedCount >= pageSize)
-                                    {
-                                        keepRunning = false;
-                                        break;
-                                    }
+                                    lastEtag = referencedDocument.Etag;
+                                    inMemoryStats.UpdateLastEtag(lastEtag, isTombstone);
 
-                                    if (MapDocuments.MaybeRenewTransaction(queryContext, sw, _configuration, ref maxTimeForDocumentTransactionToRemainOpen))
+                                    if (CanContinueReferenceBatch() == false)
                                         break;
                                 }
 
                                 if (hasChanges == false)
                                     break;
                             }
+
+                            bool CanContinueReferenceBatch()
+                            {
+                                if (CanContinueBatch(queryContext, indexContext, collectionStats, indexWriter, lastEtag, lastCollectionEtag, totalProcessedCount) == false)
+                                {
+                                    keepRunning = false;
+                                    return false;
+                                }
+
+                                if (totalProcessedCount >= pageSize)
+                                {
+                                    keepRunning = false;
+                                    return false;
+                                }
+
+                                if (MapItems.MaybeRenewTransaction(queryContext, sw, _configuration, ref maxTimeForDocumentTransactionToRemainOpen))
```

Comment:
Do we want to avoid doing this for every item? Should this be also done every 128 items?

Reply:
done

Reply:
Where was this done?